lib/express/server.js

```diff
@@ -164,16 +101,131 @@ function configureServer (app, routeConfiguringFunction, allowCors = true, baseP
   app.all('/test/guinea-pig-scrollable', guineaPigScrollable);
   app.all('/test/guinea-pig-app-banner', guineaPigAppBanner);
+}
 
-  // catch this last, so anything that falls through is 404ed
-  app.use(catch404Handler);
+function configureHttp ({httpServer, reject}) {
+  const serverState = {
+    notifier: new EventEmitter(),
+    closed: false,
+  };
+  httpServer.addWebSocketHandler = addWebSocketHandler;
+  httpServer.removeWebSocketHandler = removeWebSocketHandler;
+  httpServer.removeAllWebSocketHandlers = removeAllWebSocketHandlers;
+  httpServer.getWebSocketHandlers = getWebSocketHandlers;
+
+  // http.Server.close() only stops new connections, but we need to wait until
+  // all connections are closed and the `close` event is emitted
+  const close = httpServer.close.bind(httpServer);
+  httpServer.close = async () => await new B((resolve, reject) => {
+    // https://github.com/nodejs/node-v0.x-archive/issues/9066#issuecomment-124210576
+    serverState.closed = true;
+    serverState.notifier.emit('shutdown');
+    httpServer.on('close', resolve);
+    close((err) => {
+      if (err) reject(err); // eslint-disable-line curly
+    });
+  });
+
+  httpServer.on('error', (err) => {
+    if (err.code === 'EADDRNOTAVAIL') {
+      log.error('Could not start REST http interface listener. ' +
+                'Requested address is not available.');
+    } else {
+      log.error('Could not start REST http interface listener. The requested ' +
+                'port may already be in use. Please make sure there is no ' +
+                'other instance of this server running already.');
+    }
+    reject(err);
+  });
+
+  httpServer.on('connection', (socket) => {
+    socket.setTimeout(KEEP_ALIVE_TIMEOUT_MS);
+    socket.on('error', reject);
+
+    function destroy () {
+      if (socket._openReqCount === 0) {
+        socket.destroy();
+      }
+    }
+    socket._openReqCount = 0;
+    socket.once('close', () => serverState.notifier.removeListener('shutdown', destroy));
+    serverState.notifier.once('shutdown', destroy);
+  });
+
+  httpServer.on('request', function (req, res) {
+    const socket = req.connection || req.socket;
+    socket._openReqCount++;
+    res.on('finish', function () {
+      socket._openReqCount--;
+      if (serverState.closed && socket._openReqCount === 0) {
+        socket.destroy();
+      }
+    });
+  });
+}
+
+async function configureServerPlugins ({plugins, app, httpServer}) {
+  // allow plugins to update the app and http server objects
+  for (const plugin of plugins.filter((p) => p.updatesServer)) {
+    log.info(`Allowing plugin ${plugin.name} to modify the Appium server`);
+    try {
+      // wrap this in error handling in case the plugin forgot to mark the function async or
+      // forgot to implement it
+      await plugin.updateServer(app, httpServer);
+    } catch (err) {
+      log.error(`Plugin '${plugin.name}' tried to update the server but the updateServer ` +
```

Comment:
it makes sense to write the original error stack into debug log here

Reply:
stack will be printed when the original error is thrown in the main appium process